src/providers/aws/mod.rs

```diff
@@ -49,29 +50,74 @@ pub struct AwsProvider {
 
 impl AwsProvider {
     pub fn try_new() -> Result<AwsProvider> {
-        let client = retry::Client::try_new()?.return_on_404(true);
+        let mut client = retry::Client::try_new()?.return_on_404(true);
+        let token = AwsProvider::fetch_imdsv2_token(client.clone());
+
+        // If IMDSv2 token is fetched successfully, set the header.
+        // Otherwise, proceed with IMDSv1 mechanism.
+        match token {
+            Ok(t) => {
+                client = client.header(
+                    header::HeaderName::from_bytes(b"X-aws-ec2-metadata-token")
+                        .chain_err(|| "setting header name for aws imdsv2 metadata")?,
+                    header::HeaderValue::from_bytes(t.as_bytes())
+                        .chain_err(|| "setting header value for aws imdsv2 metadata")?,
+                );
+            }
+            Err(err) => {
+                warn!("failed to fetch aws imdsv2 session token with: {}", err);
+            }
+        }
 
         Ok(AwsProvider { client })
     }
 
     #[cfg(test)]
-    fn endpoint_for(key: &str) -> String {
+    #[allow(unused_variables)]
+    fn endpoint_for(key: &str, use_latest: bool) -> String {
         let url = mockito::server_url();
         format!("{}/{}", url, key)
     }
 
     #[cfg(not(test))]
-    fn endpoint_for(key: &str) -> String {
+    fn endpoint_for(key: &str, use_latest: bool) -> String {
         const URL: &str = "http://169.254.169.254/2019-10-01";
-        format!("{}/{}", URL, key)
+        const URL_LATEST: &str = "http://169.254.169.254/latest";
+        if use_latest {
+            format!("{}/{}", URL_LATEST, key)
+        } else {
+            format!("{}/{}", URL, key)
+        }
+    }
+
+    fn fetch_imdsv2_token(client: retry::Client) -> Result<String> {
+        let token: String = client
+            .header(
+                header::HeaderName::from_bytes(b"X-aws-ec2-metadata-token-ttl-seconds")
+                    .chain_err(|| "setting header name for aws imdsv2 token")?,
+                header::HeaderValue::from_bytes(b"21600")
+                    .chain_err(|| "setting header value for aws imdsv2 token")?,
+            )
+            .put(
+                retry::Raw,
+                #[cfg(test)]
+                AwsProvider::endpoint_for("api/token", false),
```

Comment:
s/false/true/

But you can drop these two `cfg[]` branches here completely now, as you already moved the conditional part to `endpoint_for` declaration.